src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -87,18 +88,39 @@ internal static class DeserializationHelpers
             }
         }
 
-        internal static void ApplyInstanceAnnotations(object resource, IEdmStructuredTypeReference structuredType, ODataResource oDataResource,
+        internal static void ApplyInstanceAnnotations(object resource, IEdmStructuredTypeReference structuredType, ODataResourceBase oDataResource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
         {
-            PropertyInfo propertyInfo = EdmLibHelpers.GetInstanceAnnotationsContainer(structuredType.StructuredDefinition(), readContext.Model);
-            if (propertyInfo == null)
+            //Apply instance annotations for both entityobject/changedobject/delta and normal resources
+            IODataInstanceAnnotationContainer instanceAnnotationContainer = null;
```

Comment:
Given the introduction of `transientInstanceAnnotationContainer`, I think I'd rename `instanceAnnotationContainer` to `persistentInstanceAnnotationContainer` just to be clear.